src/atlas/projection/detail/CubedSphereProjectionBase.cc

```diff
@@ -334,6 +295,146 @@ void CubedSphereProjectionBase::schmidtTransform( double stretchFac, double targ
 
 }
 
+idx_t CubedSphereProjectionBase::tileFromXY( const double xy[] ) const {
+  // Assume one face-edge is of length 90 degrees.
+  //
+  //   y ^
+  //     |
+  //    135              ----------
+  //     |              |     ^    |
+  //     |              |          |
+  //     |              |=<   2   <|
+  //     |              |     v    |
+  //     |              |     =    |
+  //     45  0----------2----------3----------4----------
+  //     |   |    ^     |     ^    |    =     |     =    |
+  //     |   |          |          |    ^     |     ^    |
+  //     |   |=<  0    <|=<   1   <|=<  3    <|=<   4   <|
+  //     |   |    v     |     v    |          |          |
+  //     |   |    =     |     =    |    v     |     v    |
+  //    -45  0 ---------1----------1----------5----------
+  //     |                                    |     =    |
+  //     |                                    |     ^    |
+  //     |                                    |=<   5   <|
+  //     |                                    |          |
+  //     |                                    |     v    |
+  //   -135                                    ----------(5 for end iterator)
+  //     ----0---------90--------180--------270--------360--->  x
+
+  idx_t t{-1};
+
+  if ((xy[LON] >= 0.) && ( xy[LAT] >= -45.) && (xy[LON] < 90.) && (xy[LAT] < 45.)) {
+     t = 0;
+  } else if ((xy[LON] >= 90.) && ( xy[LAT] >= -45.) && (xy[LON] < 180.) && (xy[LAT] < 45.)) {
+     t = 1;
+  } else if ((xy[LON] >= 90.) && ( xy[LAT] >= 45.) && (xy[LON] < 180.) && (xy[LAT] < 135.)) {
+     t = 2;
+  } else if ((xy[LON] >= 180.) && ( xy[LAT] > -45.) && (xy[LON] < 270.) && (xy[LAT] <= 45.)) {
+     t = 3;
+  } else if ((xy[LON] >= 270.) && ( xy[LAT] > -45.) && (xy[LON] < 360.) && (xy[LAT] <= 45.)) {
+     t = 4;
+  } else if ((xy[LON] >= 270.) && ( xy[LAT] > -135.) && (xy[LON] < 360.) && (xy[LAT] <= -45.)) {
+     t = 5;
+  }
+
+  // extra points
+  if ((std::abs(xy[LON]) < 1e-13) && (std::abs(xy[LAT] - 45.) < 1e-13)) t = 0;
+  if ((std::abs(xy[LON] - 180.) < 1e-13) && (std::abs(xy[LAT] + 45.) < 1e-13)) t = 1;
+
+  // for end iterator !!!!
+  if ((std::abs(xy[LON] - 360.) < 1e-13) && (std::abs(xy[LAT] + 135.) < 1e-13)) t = 5;
+
+  return t;
+}
+
+// assuming that crd[] here holds the latitude/longitude in RADIANS.
+// expecting longitude range between 0 and 2* PI
+idx_t CubedSphereProjectionBase::tileFromLonLat(const double crd[]) const {
+    idx_t t(-1);
+    double xyz[3];
+
```

Comment:
I still think this calculation is more complicated than it needs to be.

You can just define a unit cube centred at zero with six `PointXYZ` objects (`p0`) defining the face centres and six more defining the normal vectors (`n`) of each face.

You then just calculate
```
d = p0 dot n / ( xyz dot n)
```
for each face and the smallest +ve `d` corresponds to the right tile. You can add in some safety code to check for div-zero, but even without it the IEEE 754 standard should be able to cope with +/- inf.

Reply:
I honestly don't think that the current calculation is that complicated. All the complexity comes from the corner and edges cases not from the general algorithm,  ( ... I mean literally the corners and edges of the cubed which happen to be much more complicated than what is in LFRic).

1.

Reply:
Yeah, there's no point re-writing it. It clearly works, I'd have just done it differently.